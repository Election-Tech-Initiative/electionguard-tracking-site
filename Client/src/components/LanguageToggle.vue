<template>
    <div class="language-toggle">
        <span class="language-label">
            {{ $t("languageLabel") }}
        </span>
        <div class="locale-changer select is-small">
            <select v-model="selectedLanguage" @change="onLanguageChange">
                <option v-for="(lang, i) in supportedLanguages" :key="`lang${i}`" :value="lang[0]">{{ lang[1] }}</option>
            </select>
        </div>
    </div>
</template>

<script lang="ts">
import { Component, Vue, Watch } from 'vue-property-decorator';

import { getSupportedLanguages } from '@/i18n';

@Component
export default class LanguageToggle extends Vue {
  public supportedLanguages = getSupportedLanguages();
  public selectedLanguage: string | undefined;

  @Watch('$i18n.locale')
  public onLocaleChange(value: string, oldValue: string) {
      if (value !== oldValue) {
        this.selectedLanguage = value;
      }
  }

  public created(): void {
      this.selectedLanguage = this.$i18n.locale;
  }

  public onLanguageChange(): void {
    const query = `?lang=${this.selectedLanguage}`;
    this.$router.push(`${this.$route.path}${query}`);
  }
}
</script>

<style lang="scss" scoped>
    .language-label {
        color: $dark;
        font-weight: 600;
    }
    .locale-changer {
        vertical-align: middle;
    }
</style>