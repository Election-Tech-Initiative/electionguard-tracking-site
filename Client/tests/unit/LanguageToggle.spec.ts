import { mount } from '@vue/test-utils';

import LanguageToggle from '@/components/LanguageToggle.vue';

jest.mock('@/i18n', () => ({ getSupportedLanguages: () => [['en', 'English'], ['es', 'Spanish']] }));

describe('LanguageToggle.vue', () => {
    const mockedPath = 'mocked.path';
    const mockRouterPush = jest.fn();
    const wrapper = mount(LanguageToggle, {
        mocks: {
            $i18n: {
                locale: 'en',
            },
            $route: {
                path: mockedPath,
            },
            $router: {
                push: mockRouterPush,
            },
            $t: (msg: string): string => msg
        },
    });

    it('renders', () => {
        expect(wrapper.element).toMatchSnapshot();
    });

    it('changing language pushes new route with lang query string', () => {
        const lang = 'es';
        const select = wrapper.find('select');
        select.setValue(lang);
        expect(mockRouterPush).toBeCalledWith(`${mockedPath}?lang=${lang}`);
    });

    it('$i18n.locale change updates selectedLanguage', () => {
        wrapper.setData({ $i18n: { locale: 'fr' }});
        expect((wrapper.vm as any).selectedLanguage).toBe('fr');
    });
});
