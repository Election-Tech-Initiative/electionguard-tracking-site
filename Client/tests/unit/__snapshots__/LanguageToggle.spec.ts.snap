// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguageToggle.vue renders 1`] = `
<div
  class="language-toggle"
>
  <span
    class="language-label"
  >
    
        languageLabel
    
  </span>
   
  <div
    class="locale-changer select is-small"
  >
    <select>
      <option
        value="en"
      >
        English
      </option>
      <option
        value="es"
      >
        Spanish
      </option>
    </select>
  </div>
</div>
`;
